[PyTorchEdge] Making promoted ops to instructions

pavithranrao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70610

Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

Commit:
[PyTorchEdge] Making promoted ops to instructions



Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

zhxchen17 commented on torch/csrc/jit/mobile/promoted_prim_ops.cpp:
```diff
@@ -99,15 +100,15 @@ void toList(Stack& stack) {
+  at::TypePtr out_ty;
   if (elem_ty_val == 0) {
-    out_ty = IntType::get();
+    out_ty = at::IntType::get();
```
Since these operators become part of the core interpreter, we should use DynamicType from now.
For example, `IntType::get()` should become `DynamicTypeTrait<IntType>::getBaseType()`. If that looks tedious feel free to write some helper functions.

zhxchen17 commented on torch/csrc/jit/mobile/promoted_prim_ops.cpp:
```diff
@@ -120,8 +121,8 @@ void toList(Stack& stack) {
   TORCH_CHECK(
-      (out_ty == FloatType::get() && t.is_floating_point()) ||
-          (out_ty == ComplexType::get() && t.is_complex()) ||
+      (out_ty == at::FloatType::get() && t.is_floating_point()) ||
```
Same. We should be using DynamicType for saving binary size.

zhxchen17 commented on torch/csrc/jit/mobile/promoted_prim_ops.cpp:
```diff
@@ -134,7 +135,7 @@ void toList(Stack& stack) {
   for (const auto i : c10::irange(dim_val)) {
     (void)i; // Suppress unused variable warning
-    out_ty = ListType::create(out_ty);
+    out_ty = at::ListType::create(out_ty);
```
`DynamicTypeFactory::create<ListType>(out_ty);`
DynamicTypeFactory's header is ATen/core/type_factory.h

zhxchen17 commented on torch/csrc/jit/runtime/register_ops_common_utils.cpp:
```diff
@@ -0,0 +1,119 @@
+#include <torch/csrc/jit/runtime/register_ops_common_utils.h>
```
I feel all the helper functions in this file are used by mobile side, so maybe it's better to move this and the header under `jit/mobile` folder?

zhxchen17 commented on torch/csrc/jit/runtime/register_ops_common_utils.cpp:
```diff
@@ -0,0 +1,119 @@
+    at::IntArrayRef strides,
+    size_t element_size) {
+  // If ty is a ListType, get the element type.
+  if (auto list_type = ty->cast<at::ListType>()) {
```
same to toList().

zhxchen17 commented on torch/csrc/jit/runtime/register_ops_common_utils.h:
```diff
@@ -0,0 +1,35 @@
+    char* data,
+    int64_t cur_dim,
+    int64_t num_tensor_dims,
+    at::TypePtr ty,
```
We probably want to make the signature to be DynamicTypePtr here to hint the API usage.

zhxchen17 reviewed (changes requested):
this is pretty cool :) some comments.

pavithranrao committed:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

pavithranrao committed:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

pavithranrao committed:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

pavithranrao committed:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

zhxchen17 commented:
@pavithranrao just fyi this week I add some helper function to the dynamic type API. Now we can do something like this:
```
#include <ATen/core/type_factory.h>
c10::dynT<IntType>(); // instead of IntType::get();
c10::dynT<ListType>(c10::dynT<IntType>()); // instead ListType::create(IntType::get());
```

iseeyuan commented on torch/csrc/jit/mobile/register_ops_common_utils.cpp:
```diff
@@ -0,0 +1,122 @@
+  return idx;
+}
+
+C10_UNUSED at::Tensor to_dispatch(
```
Any reason to move this function in a common file? I couldn't find the new usage.

iseeyuan reviewed (changes requested):
1. Could you confirm if RAISE_EXCEPTION are calling the correct function? 
2. Could you change to dynamic type per @zhxchen17 's comments?

pavithranrao commented on test/cpp/jit/test_backend.cpp:
```diff
@@ -187,6 +187,28 @@ TEST(BackendTest, TestComposite) {
   AT_ASSERT(res_jit.toTensor().equal(res_mobile.toTensor()));
 }
 
+TEST(BackendTest, TestPrimDtype) {
```
will remove

pavithranrao commented on torch/csrc/jit/mobile/interpreter.cpp:
```diff
@@ -248,6 +248,63 @@ bool InterpreterState::run(Stack& stack) {
+        } break;
+        case RAISE_EXCEPTION: {
+          raiseException(stack);
+          // raiseExceptionWithMessage(stack);
```
I am thinking of pushing this PR without the two arguments. Once it lands, we bump version and in Shunting's diff we can enable this kernel.

iseeyuan replied:
The consumer side must be consistent between this PR and version bump. If this PR is landed with one arg, and only changed to 2-arg when the version is bumped, we still have 1-arg consumer in the wild, and there can be breakage. If the new producer is disabled by default, there should be no RAISE_EXCEPTION in models in the wild. Any risk I missed?

pavithranrao replied:
oh makes sense, will rebase this on top of Shunting's PR

pavithranrao committed:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

pavithranrao committed:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

pavithranrao committed:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

iseeyuan commented:
@pavithranrao I have added "ciflow/all" in labels. Please add the same label to the version bump PR to have all CI tests.

pavithranrao committed:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]

iseeyuan reviewed (approved):
LGTM. Please remember to use raiseExceptionWithMessage before landing this diff.

pavithranrao committed:
Update on "[PyTorchEdge] Making promoted ops to instructions"


Differential Revision: [D33398092](https://our.internmc.facebook.com/intern/diff/D33398092/)

[ghstack-poisoned]
